File: src/accessiweather/weather_anomaly.py
"""
Historical temperature anomaly detection for AccessiWeather.

Computes a multi-year baseline temperature for a given date and compares
current conditions against it to produce concise anomaly callouts.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass
from datetime import date, timedelta
from typing import TYPE_CHECKING, Literal

if TYPE_CHECKING:
    from .openmeteo_client import OpenMeteoApiClient

logger = logging.getLogger(__name__)

YEARS_OF_HISTORY = 5
MIN_YEARS_REQUIRED = 3
DATE_WINDOW_DAYS = 7


@dataclass
class AnomalyCallout:
    """Describes how current temperature compares to the historical baseline."""

    temp_anomaly: float  # positive = warmer, negative = cooler (°F)
    temp_anomaly_description: str  # human-readable line
    precip_anomaly_description: str | None
    severity: Literal["normal", "notable", "significant"]


def _classify_severity(anomaly_f: float) -> Literal["normal", "notable", "significant"]:
    """Classify anomaly magnitude into severity buckets."""
    abs_diff = abs(anomaly_f)
    if abs_diff >= 5.0:
        return "significant"
    if abs_diff >= 2.0:
        return "notable"
    return "normal"


def _build_description(anomaly_f: float, years: int) -> str:
    """Build a human-readable anomaly description."""
    abs_diff = abs(anomaly_f)
    direction = "warmer" if anomaly_f > 0 else "cooler"
    if abs_diff < 0.5:
        return f"Near the {years}-year average for this date."
    return (
        f"Currently {abs_diff:.1f}\u00b0F {direction} than the {years}-year average for this date."
    )


def compute_anomaly(
    lat: float,
    lon: float,
    current_temp_f: float,
    current_date: date,
    client: OpenMeteoApiClient,
) -> AnomalyCallout | None:
    """
    Compute historical temperature anomaly for the given location and date.

    Fetches the last YEARS_OF_HISTORY years of daily temperature data for a
    +/-DATE_WINDOW_DAYS window around current_date and computes the baseline mean.

    Returns None if fewer than MIN_YEARS_REQUIRED years of data are available
    or if all fetches fail.
    """
    yearly_means: list[float] = []

    for years_back in range(1, YEARS_OF_HISTORY + 1):
        try:
            anchor = current_date.replace(year=current_date.year - years_back)
        except ValueError:
            # Feb 29 on a non-leap year
            anchor = current_date.replace(year=current_date.year - years_back, day=28)

        start = anchor - timedelta(days=DATE_WINDOW_DAYS)
        end = anchor + timedelta(days=DATE_WINDOW_DAYS)

        # Archive API lags ~5 days behind present; clamp to avoid requesting
        # dates that are too recent to have archive data.
        too_recent = date.today() - timedelta(days=5)
        if end > too_recent:
            end = too_recent
        if start >= end:
            logger.debug("Skipping year -%d: date window not yet archived", years_back)
            continue

        try:
            params = {
                "latitude": lat,
                "longitude": lon,
                "start_date": start.isoformat(),
                "end_date": end.isoformat(),
                "daily": ["temperature_2m_mean"],
                "temperature_unit": "fahrenheit",
                "timezone": "auto",
            }
            response = client._make_request("archive", params, use_archive=True)
        except Exception as exc:
            logger.debug("Anomaly archive fetch failed for year -%d: %s", years_back, exc)
            continue

        if not response or "daily" not in response:
            continue

        temps = response["daily"].get("temperature_2m_mean") or []
        valid = [t for t in temps if t is not None]
        if not valid:
            continue

        yearly_means.append(sum(valid) / len(valid))

    if len(yearly_means) < MIN_YEARS_REQUIRED:
        logger.debug(
            "Insufficient historical data: %d years (need %d)",
            len(yearly_means),
            MIN_YEARS_REQUIRED,
        )
        return None

    baseline = sum(yearly_means) / len(yearly_means)
    anomaly = current_temp_f - baseline
    severity = _classify_severity(anomaly)
    description = _build_description(anomaly, len(yearly_means))

    return AnomalyCallout(
        temp_anomaly=anomaly,
        temp_anomaly_description=description,
        precip_anomaly_description=None,
        severity=severity,
    )

File: rust/crates/aw-core/src/weather_anomaly.rs
//! Historical temperature anomaly callouts. Port of
//! `accessiweather/weather_anomaly.py`; the archive requests are made by the
//! caller-supplied `fetch` closure (see `aw_providers::client::history`).

use chrono::{Datelike, Duration, NaiveDate};
use serde_json::Value;

use crate::model::AnomalyCallout;

pub const YEARS_OF_HISTORY: i32 = 5;
pub const MIN_YEARS_REQUIRED: usize = 3;
pub const DATE_WINDOW_DAYS: i64 = 7;

pub fn classify_severity(anomaly_f: f64) -> &'static str {
    let diff = anomaly_f.abs();
    if diff >= 5.0 {
        "significant"
    } else if diff >= 2.0 {
        "notable"
    } else {
        "normal"
    }
}

pub fn build_description(anomaly_f: f64, years: usize) -> String {
    let diff = anomaly_f.abs();
    if diff < 0.5 {
        return format!("Near the {years}-year average for this date.");
    }
    let direction = if anomaly_f > 0.0 { "warmer" } else { "cooler" };
    format!("Currently {diff:.1}\u{b0}F {direction} than the {years}-year average for this date.")
}

/// `compute_anomaly`: compare `current_temp_f` with the mean daily
/// temperature of the same ±7-day window in each of the last five years.
///
/// `fetch(start, end)` returns the archive JSON for a date range
/// (`daily=temperature_2m_mean`, Fahrenheit), or `None` on failure. `today`
/// clamps windows the archive (≈5 days behind) cannot serve yet.
pub fn compute_anomaly(
    current_temp_f: f64,
    current_date: NaiveDate,
    today: NaiveDate,
    mut fetch: impl FnMut(NaiveDate, NaiveDate) -> Option<Value>,
) -> Option<AnomalyCallout> {
    let mut yearly_means = Vec::new();
    for years_back in 1..=YEARS_OF_HISTORY {
        let year = current_date.year() - years_back;
        let anchor = current_date
            .with_year(year)
            .or_else(|| NaiveDate::from_ymd_opt(year, current_date.month(), 28))?;
        let start = anchor - Duration::days(DATE_WINDOW_DAYS);
        let end = (anchor + Duration::days(DATE_WINDOW_DAYS)).min(today - Duration::days(5));
        if start >= end {
            continue;
        }
        let Some(response) = fetch(start, end) else {
            continue;
        };
        let temps: Vec<f64> = response
            .get("daily")
            .and_then(|d| d.get("temperature_2m_mean"))
            .and_then(Value::as_array)
            .map(|values| values.iter().filter_map(Value::as_f64).collect())
            .unwrap_or_default();
        if temps.is_empty() {
            continue;
        }
        yearly_means.push(temps.iter().sum::<f64>() / temps.len() as f64);
    }
    if yearly_means.len() < MIN_YEARS_REQUIRED {
        return None;
    }
    let baseline = yearly_means.iter().sum::<f64>() / yearly_means.len() as f64;
    let anomaly = current_temp_f - baseline;
    Some(AnomalyCallout {
        temp_anomaly: anomaly,
        temp_anomaly_description: build_description(anomaly, yearly_means.len()),
        precip_anomaly_description: None,
        severity: classify_severity(anomaly).into(),
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::golden::{self, field};

    #[test]
    fn golden_anomalies() {
        let cases = golden::load("history/anomaly.json");
        for case in cases.as_array().unwrap() {
            let responses = case["responses"].as_object().unwrap().clone();
            let mut requests = Vec::new();
            let result = compute_anomaly(
                field(case, "current_temp_f"),
                field(case, "current_date"),
                field(case, "today"),
                |start, end| {
                    let key = format!("{start}/{end}");
                    requests.push(key.clone());
                    responses.get(&key).cloned().filter(|v| !v.is_null())
                },
            );
            let expected: Option<AnomalyCallout> = field(case, "result");
            let expected_requests: Vec<String> = field(case, "requests");
            assert_eq!(result, expected, "{}", case["name"]);
            assert_eq!(requests, expected_requests, "{}", case["name"]);
        }
    }

    #[test]
    fn severity_buckets_and_text() {
        assert_eq!(classify_severity(-5.0), "significant");
        assert_eq!(classify_severity(2.0), "notable");
        assert_eq!(classify_severity(1.9), "normal");
        assert_eq!(
            build_description(0.4, 5),
            "Near the 5-year average for this date."
        );
        assert_eq!(
            build_description(-3.25, 4),
            "Currently 3.2°F cooler than the 4-year average for this date."
        );
    }
}
